common/src/IO/ObjParser.cpp

```diff
@@ -0,0 +1,250 @@
+ Copyright (C) 2010-2017 Kristian Duske
+
+ This file is part of TrenchBroom.
+
+ TrenchBroom is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ TrenchBroom is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with TrenchBroom. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ObjParser.h"
+
+#include "Logger.h"
+#include "Assets/EntityModel.h"
+#include "Assets/Texture.h"
+#include "IO/FileSystem.h"
+#include "IO/FreeImageTextureReader.h"
+#include "IO/Path.h"
+#include "Renderer/PrimType.h"
+#include "Renderer/TexturedIndexRangeMap.h"
+#include "Renderer/TexturedIndexRangeMapBuilder.h"
+
+#include <vecmath/forward.h>
+#include <kdl/string_utils.h>
+
+#include <functional>
+#include <string>
+
+namespace TrenchBroom {
+    namespace IO {
+
+        struct ObjVertexRef {
+            /**
+             * Parses a vertex reference.
+             *
+             * @param text the text of this reference (such as "1/2/3")
+             */
+            ObjVertexRef(const std::string& text) : m_texcoord(0) {
+                const auto components = kdl::str_split(text, "/");
+                std::vector<size_t> components_num;
+                for (const std::string& com : components) {
+                    const auto csize = kdl::str_to_size(com);
+                    if (csize) {
+                        components_num.push_back(*csize);
+                    } else {
+                        throw ParserException("OBJ file has invalid number in vertex reference");
+                    }
+                }
+                if (components_num.size() == 0) {
+                    throw ParserException("OBJ file has an empty vertex reference");
+                } else if (components_num.size() == 1) {
+                    m_position = components_num[0];
+                } else {
+                    m_position = components_num[0];
+                    m_texcoord = components_num[1];
+                }
+            }
+
+            // Position index (1-based. Should always be present.)
+            size_t m_position;
+            // Texture coordinate index (Also 1-based. If not present, 0.)
+            size_t m_texcoord;
+        };
+
+        struct ObjFace {
+            // The material of this face (as a skin index)
+            size_t m_material;
+            // The vertices of this face.
+            std::vector<ObjVertexRef> m_vertices;
+        };
+
+        ObjParser::ObjParser(const std::string& name, const char* begin, const char* end) :
+        m_name(name), m_text(begin, (size_t) (end - begin)) {}
```

Comment:
Sorry, missed one here: We don't use C-style casts. In this case you don't need to cast at all, just pass the char pointers. `std::string` has a constructor that takes two iterators from which to create the string.

Reply:
Resolved as of 5993f765ab.